pennylane/resources/factorization.py

```diff
@@ -0,0 +1,136 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the functions needed for two-electron tensor factorization.
+"""
+from pennylane import numpy as np
+
+
+def factorize(two, tol):
+    r"""Return the double-factorized form of a two-electron tensor.
+
+    The second quantized electronic Hamiltonian is constructed in terms of fermionic creation,
+    :math:`a^{\dagger}` , and annihilation, :math:`a`, operators as
+    [`arXiv:1902.02134 <https://arxiv.org/abs/1902.02134>`_]
+
+    .. math::
+
+        H = \sum_{\alpha \in \{\uparrow, \downarrow \} } \sum_{pq} h_{pq} a_{p,\alpha}^{\dagger}
+        a_{q, \alpha} + \frac{1}{2} \sum_{\alpha, \beta \in \{\uparrow, \downarrow \} } \sum_{pqrs}
```

Comment:
@ixfoduap @soranjh something worth thinking about: this is a lot of background/math to scroll through before you get to the example and arguments!

I tend to be of the opinion that we should structure our docstrings for _information_ first (think: reference-style), and _understanding_ second (how does it work internally, what is the underlying theory). The thinking is, this docstring will be used by both new users trying to understand it, but also by _repeated_ users who will need to look up the args/examples for reference, and we should make this easy to find.

[Of course, just roughly -- the first summary paragraph should try to provide high-level details of both].

In this case, I'm wondering if we should do this:

- Summary paragraph giving overview of what this function does/when you would use it.
- Args
- Returns
- Example
- Theory

Now that we have generalized `.. details::` section, we also have the ability to do:

```rest
.. details::
  :title: Theory
```

Like how we do with more advanced usage details :)

Reply:
NumPy takes a similar approach, although swaps the Theory and Example sections: https://numpy.org/doc/stable/reference/generated/numpy.linalg.cholesky.html#numpy.linalg.cholesky.

Although, the theory in NumPy is often less involved than in quantum!